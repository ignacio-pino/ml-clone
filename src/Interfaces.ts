export interface IProductListData {
  id: string;
  title: string;
  price: string;
  original_price?: string;
  currency_id: "UYU" | "USD";
  thumbnail: string;
  shipping: {
    free_shipping: boolean;
  };
}

export interface ISortDataList {
  sort: ISortData;
  available_sorts: ISortData[];
}

export interface ISortData {
  id: string;
  name: string;
}
export interface IProductCard {
  id: string;
  title: string;
  price: string;
  original_price?: string;
  currency_id: "UYU" | "USD";
  thumbnail: string;

  free_shipping: boolean;
}

export interface IFilterDataList {
  filters: IFilterList[];
  availableFilters: IFilterList[];
}
export interface IFilterList {
  id: string;
  name: string;
  type: string;
  values: IFilterListValues[];
}

export interface IFilterListValues {
  id: string;
  name: string;
  results?: number;
  path_from_root?: IPathFromRoot[];
}

export interface ICategoryData {
  query?: string;
  categoryInformation?: ICategoryInformation;
  totalResults: number;
  childrenCategories?: IPathFromRoot[];
}

export interface ICategoryInformation {
  id: string;
  name: string;
  path_from_root?: IPathFromRoot[];
}

export interface IPathFromRoot {
  id: string;
  name: string;
}

export interface IChildrenCategoryData {
  id: string;
  name: string;
  total_items_in_this_category: number;
}

export interface IHeaderCategoryData {
  categoryId: string;
  categoryName: string;
}
