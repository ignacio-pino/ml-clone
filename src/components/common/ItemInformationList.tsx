import { HStack, VStack, Text, Divider, Button } from "@chakra-ui/react";
import { IItemInformationList } from "../../Interfaces";
import { EmailIcon, InfoIcon, LockIcon, RepeatIcon } from "@chakra-ui/icons";

export default function ItemInformationList({
  itemInformationList,
}: {
  itemInformationList: IItemInformationList;
}) {
  const {
    condition,
    sold_quantity,
    title,
    original_price,
    price,
    currency_id,
    shipping,
    available_quantity,
  } = itemInformationList;

  const soldQuantityReference = [
    { min: 6, max: 25, result: 5 },
    { min: 26, max: 50, result: 25 },
    { min: 51, max: 100, result: 50 },
    { min: 101, max: 150, result: 100 },
    { min: 151, max: 200, result: 150 },
    { min: 201, max: 250, result: 200 },
    { min: 251, max: 500, result: 250 },
    { min: 501, max: 5000, result: 500 },
    { min: 5001, max: 50000, result: 5000 },
    { min: 50001, max: 500000, result: 50000 },
  ];

  const availableQuantityReference = [
    { min: 51, max: 100, result: 50 },
    { min: 101, max: 150, result: 100 },
    { min: 151, max: 200, result: 150 },
    { min: 201, max: 250, result: 200 },
    { min: 251, max: 500, result: 250 },
    { min: 501, max: 5000, result: 500 },
    { min: 5001, max: 50000, result: 5000 },
    { min: 50001, max: 99999, result: 50000 },
  ];

  const convertToRange = ({
    value,
    reference,
  }: {
    value: number;
    reference: { min: number; max: number; result: number }[];
  }): number | null => {
    const matchingRange = reference.find(
      ({ min, max }) => value >= min && value <= max
    );

    return matchingRange ? matchingRange.result : null;
  };

  return (
    <VStack
      alignItems={"flex-start"}
      border={"1px"}
      borderColor={"gray.500"}
      borderRadius={"10px"}
      p={6}
    >
      <HStack>
        <Text color={"gray.500"} fontSize={"sm"}>
          {condition === "new" ? "Nuevo" : "Usado"}
        </Text>
        <Divider orientation="vertical" color={"gray.500"} height={"1rem"} />
        <Text color={"gray.500"} fontSize={"sm"}>
          {sold_quantity > 5
            ? `+${convertToRange({
                value: sold_quantity,
                reference: soldQuantityReference,
              })}`
            : sold_quantity}{" "}
          vendidos
        </Text>
      </HStack>
      <Text as="b" fontSize={"lg"}>
        {title}
      </Text>
      <VStack alignItems={"flex-start"}>
        {original_price && (
          <HStack>
            <Text textColor={"gray.500"} as="del" fontSize={"xs"}>
              {currency_id === "USD" ? "U$S" : "$"} {original_price}
            </Text>
          </HStack>
        )}
        <HStack mb="2" mt={original_price ? "-3" : "5"}>
          <HStack gap={0.5}>
            <Text as="b" fontSize={"x-large"}>
              {currency_id === "USD" ? "U$S" : "$"} {price}
            </Text>
          </HStack>

          {original_price && (
            <Text textColor={"purple.400"}>
              {Math.floor(((original_price - price) / original_price) * 100)}%
              OFF
            </Text>
          )}
        </HStack>
      </VStack>
      <HStack>
        <EmailIcon />
        <Text color={"purple.400"}>
          Envío:{" "}
          {shipping.free_shipping ? (
            <Text color={"white"} as={"span"}>
              Llega gratis entre el <b> martes</b> y el <b>jueves</b>
            </Text>
          ) : (
            <Text color={"white"} as={"span"}>
              Llega entre el <b> martes</b> y el <b>jueves</b> por $159{" "}
            </Text>
          )}
        </Text>
      </HStack>
      <HStack>
        <InfoIcon />
        <Text>
          {available_quantity > 50
            ? `+${convertToRange({
                value: available_quantity,
                reference: availableQuantityReference,
              })}`
            : available_quantity}{" "}
          disponibles
        </Text>
      </HStack>
      <Button backgroundColor={"purple.400"} w={"100%"} mt={"5vh"}>
        Comprar ahora
      </Button>
      <Button w={"100%"}>Añadir al carrito</Button>
      <HStack>
        <RepeatIcon />
        <Text fontSize={"sm"}>
          <Text
            fontSize={"sm"}
            as={"span"}
            color={"purple.400"}
            whiteSpace={"nowrap"}
          >
            Devolución gratis.
          </Text>{" "}
          Tienes 30 días desde que lo recibes.
        </Text>
      </HStack>
      <HStack alignItems={"flex-start"}>
        <LockIcon />
        <Text fontSize={"sm"}>
          <Text
            whiteSpace={"nowrap"}
            as={"span"}
            fontSize={"sm"}
            color={"purple.400"}
          >
            Compra Protegida,
          </Text>{" "}
          recibe el producto que esperabas o te devolvemos tu dinero.
        </Text>
      </HStack>
    </VStack>
  );
}
