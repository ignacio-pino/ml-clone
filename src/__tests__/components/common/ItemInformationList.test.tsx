import { render, screen } from "@testing-library/react";
import ItemInformationList from "../../../components/common/ItemInformationList";
import { IItemInformationList } from "../../../Interfaces";

describe("ItemInformationList Component", () => {
  const sampleItemInformationList: IItemInformationList = {
    condition: "new",
    sold_quantity: 10,
    title: "Sample Item",
    original_price: 120,
    price: 100,
    currency_id: "UYU",
    shipping: {
      free_shipping: true,
    },
    available_quantity: 60,
  };

  it("should match snapshot", () => {
    const container = render(
      <ItemInformationList itemInformationList={sampleItemInformationList} />
    ).container;

    expect(container).toMatchSnapshot();
  });

  it("renders item details correctly", () => {
    render(
      <ItemInformationList itemInformationList={sampleItemInformationList} />
    );

    expect(screen.getByText("Nuevo")).toBeInTheDocument();
    expect(screen.getByText("+5 vendidos")).toBeInTheDocument();
    expect(screen.getByText("Sample Item")).toBeInTheDocument();
    expect(screen.getByText("$ 100")).toBeInTheDocument();
    expect(screen.getByText("$ 120")).toBeInTheDocument();
    expect(screen.getByText("16% OFF")).toBeInTheDocument();
    const shippingText = screen.getByText(/Llega gratis entre el/i);
    expect(shippingText).toBeInTheDocument();

    const boldTexts = shippingText.querySelectorAll("b");
    expect(boldTexts[0].textContent).toBe(" martes");
    expect(boldTexts[1].textContent).toBe("jueves");
    expect(screen.getByText("+50 disponibles")).toBeInTheDocument();
  });

  it("renders condition as 'Usado' when condition is not 'new'", () => {
    const newItemInfo: IItemInformationList = {
      ...sampleItemInformationList,
      condition: "used",
    };
    render(<ItemInformationList itemInformationList={newItemInfo} />);

    expect(screen.getByText("Usado")).toBeInTheDocument();
  });

  it("does not render original price or discount when original_price is not provided", () => {
    const newItemInfo: IItemInformationList = {
      ...sampleItemInformationList,
      original_price: 0,
    };
    render(<ItemInformationList itemInformationList={newItemInfo} />);

    expect(screen.queryByText("U$S 120")).not.toBeInTheDocument();
    expect(screen.queryByText("20% OFF")).not.toBeInTheDocument();
  });

  it("renders shipping cost when free_shipping is false", () => {
    const newItemInfo: IItemInformationList = {
      ...sampleItemInformationList,
      shipping: { free_shipping: false },
    };
    render(<ItemInformationList itemInformationList={newItemInfo} />);

    const shippingText = screen.getByText(/Llega entre el/i);
    expect(shippingText).toBeInTheDocument();

    const boldTexts = shippingText.querySelectorAll("b");
    expect(boldTexts[0].textContent).toBe(" martes");
    expect(boldTexts[1].textContent).toBe("jueves");
  });

  it("renders currency as U$S when currency_id is USD", () => {
    const newItemInfo: IItemInformationList = {
      ...sampleItemInformationList,
      currency_id: "USD",
    };
    render(<ItemInformationList itemInformationList={newItemInfo} />);

    expect(screen.getByText("U$S 100")).toBeInTheDocument();
  });
});
