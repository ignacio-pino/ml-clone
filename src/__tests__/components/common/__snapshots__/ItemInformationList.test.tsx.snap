// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ItemInformationList Component > should match snapshot 1`] = `
<div>
  <div
    class="chakra-stack css-1kl9z2m"
  >
    <div
      class="chakra-stack css-1igwmid"
    >
      <p
        class="chakra-text css-11ox3m3"
      >
        Nuevo
      </p>
      <hr
        aria-orientation="vertical"
        class="chakra-divider css-j9wlsm"
      />
      <p
        class="chakra-text css-11ox3m3"
      >
        +5
         
        vendidos
      </p>
    </div>
    <b
      class="chakra-text css-1ezsviu"
    >
      Sample Item
    </b>
    <div
      class="chakra-stack css-1ey9w6j"
    >
      <div
        class="chakra-stack css-1igwmid"
      >
        <del
          class="chakra-text css-135297o"
        >
          $
           
          120
        </del>
      </div>
      <div
        class="chakra-stack css-1vlw5jh"
      >
        <div
          class="chakra-stack css-u2zw5"
        >
          <b
            class="chakra-text css-1llf3l7"
          >
            $
             
            100
          </b>
        </div>
        <p
          class="chakra-text css-irhq0v"
        >
          16
          % OFF
        </p>
      </div>
    </div>
    <div
      class="chakra-stack css-1igwmid"
    >
      <svg
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <g
          fill="currentColor"
        >
          <path
            d="M11.114,14.556a1.252,1.252,0,0,0,1.768,0L22.568,4.87a.5.5,0,0,0-.281-.849A1.966,1.966,0,0,0,22,4H2a1.966,1.966,0,0,0-.289.021.5.5,0,0,0-.281.849Z"
          />
          <path
            d="M23.888,5.832a.182.182,0,0,0-.2.039l-6.2,6.2a.251.251,0,0,0,0,.354l5.043,5.043a.75.75,0,1,1-1.06,1.061l-5.043-5.043a.25.25,0,0,0-.354,0l-2.129,2.129a2.75,2.75,0,0,1-3.888,0L7.926,13.488a.251.251,0,0,0-.354,0L2.529,18.531a.75.75,0,0,1-1.06-1.061l5.043-5.043a.251.251,0,0,0,0-.354l-6.2-6.2a.18.18,0,0,0-.2-.039A.182.182,0,0,0,0,6V18a2,2,0,0,0,2,2H22a2,2,0,0,0,2-2V6A.181.181,0,0,0,23.888,5.832Z"
          />
        </g>
      </svg>
      <p
        class="chakra-text css-irhq0v"
      >
        Envío:
         
        <span
          class="chakra-text css-kbn7if"
        >
          Llega gratis entre el 
          <b>
             martes
          </b>
           y el 
          <b>
            jueves
          </b>
        </span>
      </p>
    </div>
    <div
      class="chakra-stack css-1igwmid"
    >
      <svg
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12,0A12,12,0,1,0,24,12,12.013,12.013,0,0,0,12,0Zm.25,5a1.5,1.5,0,1,1-1.5,1.5A1.5,1.5,0,0,1,12.25,5ZM14.5,18.5h-4a1,1,0,0,1,0-2h.75a.25.25,0,0,0,.25-.25v-4.5a.25.25,0,0,0-.25-.25H10.5a1,1,0,0,1,0-2h1a2,2,0,0,1,2,2v4.75a.25.25,0,0,0,.25.25h.75a1,1,0,1,1,0,2Z"
          fill="currentColor"
        />
      </svg>
      <p
        class="chakra-text css-0"
      >
        +50
         
        disponibles
      </p>
    </div>
    <button
      class="chakra-button css-l3jrhf"
      type="button"
    >
      Comprar ahora
    </button>
    <button
      class="chakra-button css-uxt1e8"
      type="button"
    >
      Añadir al carrito
    </button>
    <div
      class="chakra-stack css-1igwmid"
    >
      <svg
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <g
          fill="currentColor"
        >
          <path
            d="M10.319,4.936a7.239,7.239,0,0,1,7.1,2.252,1.25,1.25,0,1,0,1.872-1.657A9.737,9.737,0,0,0,9.743,2.5,10.269,10.269,0,0,0,2.378,9.61a.249.249,0,0,1-.271.178l-1.033-.13A.491.491,0,0,0,.6,9.877a.5.5,0,0,0-.019.526l2.476,4.342a.5.5,0,0,0,.373.248.43.43,0,0,0,.062,0,.5.5,0,0,0,.359-.152l3.477-3.593a.5.5,0,0,0-.3-.844L5.15,10.172a.25.25,0,0,1-.2-.333A7.7,7.7,0,0,1,10.319,4.936Z"
          />
          <path
            d="M23.406,14.1a.5.5,0,0,0,.015-.526l-2.5-4.329A.5.5,0,0,0,20.546,9a.489.489,0,0,0-.421.151l-3.456,3.614a.5.5,0,0,0,.3.842l1.848.221a.249.249,0,0,1,.183.117.253.253,0,0,1,.023.216,7.688,7.688,0,0,1-5.369,4.9,7.243,7.243,0,0,1-7.1-2.253,1.25,1.25,0,1,0-1.872,1.656,9.74,9.74,0,0,0,9.549,3.03,10.261,10.261,0,0,0,7.369-7.12.251.251,0,0,1,.27-.179l1.058.127a.422.422,0,0,0,.06,0A.5.5,0,0,0,23.406,14.1Z"
          />
        </g>
      </svg>
      <p
        class="chakra-text css-g5hezf"
      >
        <span
          class="chakra-text css-b06hpj"
        >
          Devolución gratis.
        </span>
         
        Tienes 30 días desde que lo recibes.
      </p>
    </div>
    <div
      class="chakra-stack css-hiihwv"
    >
      <svg
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19.5,9.5h-.75V6.75a6.75,6.75,0,0,0-13.5,0V9.5H4.5a2,2,0,0,0-2,2V22a2,2,0,0,0,2,2h15a2,2,0,0,0,2-2V11.5A2,2,0,0,0,19.5,9.5Zm-9.5,6a2,2,0,1,1,3,1.723V19.5a1,1,0,0,1-2,0V17.223A1.994,1.994,0,0,1,10,15.5ZM7.75,6.75a4.25,4.25,0,0,1,8.5,0V9a.5.5,0,0,1-.5.5H8.25a.5.5,0,0,1-.5-.5Z"
          fill="currentColor"
        />
      </svg>
      <p
        class="chakra-text css-g5hezf"
      >
        <span
          class="chakra-text css-14q9065"
        >
          Compra Protegida,
        </span>
         
        recibe el producto que esperabas o te devolvemos tu dinero.
      </p>
    </div>
  </div>
</div>
`;
